### ISSUE TITLE
False positive in `identical_operands`

### ISSUE THREAD

**Issue by JaviSoto** [realm/SwiftLint#3022]
### New Issue Checklist

- [X] Updated SwiftLint to the latest version
- [X] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

The following code produces a `identical_operands` warning, but it shouldn't:

```swift
array.sorted { "\($0)" < "\($1)" }
```

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint

Done linting! Found 1 violation, 0 serious in 1276 files.
Foo.swift:23:65: warning: Identical Operands Violation: Comparing two identical operands is likely a mistake. (identical_operands)
(dummy source line)
```

### Environment

* SwiftLint version
v0.38.1
* Installation method used 
Cocoapods
* Paste your configuration file:

```yml
disabled_rules:
  - force_cast
  - notification_center_detachment
  - identifier_name
  - trailing_comma
  - type_name
  - trailing_whitespace
  - conditional_returns_on_newline
  - opening_brace
  - cyclomatic_complexity
  - todo
  - fallthrough
  - force_try
  - unneeded_parentheses_in_closure_argument
  - nesting
  - unused_closure_parameter
  - large_tuple
  - generic_type_name
  - no_fallthrough_only
opt_in_rules:
  - empty_count
  - closure_spacing
  - conditional_returns_on_newline
  - explicit_init
  - fatal_error_message
  - first_where
  - overridden_super_call
  - private_outlet
  - prohibited_super_call
  - sorted_first_last
  - override_in_extension
  - private_over_fileprivate
  - private_unit_test
  - redundant_set_access_control
  - empty_xctest_method
  - convenience_type
  - unused_import
  - unused_declaration
  - legacy_random
  - toggle_bool
  - identical_operands
  - collection_alignment
  - vertical_whitespace_opening_braces
  - vertical_whitespace_closing_braces
  - last_where
  - nslocalizedstring_key
  - contains_over_first_not_nil
  - deployment_target
  - legacy_multiple
  - optional_enum_case_matching
  - prefer_self_type_over_type_of_self
included:
  - ./
excluded:
  - <redacted>

# Rule Configuration
line_length:
  warning: 200
  ignores_interpolated_strings: true
function_body_length:
  warning: 200
  error: 400
type_body_length:
  warning: 1000
  error: 1500
file_length:
  warning: 2000
  error: 2500
function_parameter_count:
  ignores_default_parameters: true
  warning: 10
identifier_name:
  min_length:
    error: 3
  excluded:
    - id
    - URL
private_outlet:
  allow_private_set: true
overridden_super_call:
  included:
    - '*'
    - apply(_:)
unused_declaration:
  include_public_and_open: true
deployment_target:
  iOS_deployment_target: 10.0

reporter: "xcode"

# Custom Rules
# See https://github.com/realm/SwiftLint#defining-custom-rules
custom_rules:
  assert_false:
    name: "Assert False"
    regex: '(?<!\.)\b(?:Swift\.)?assert\(false,'
    message: "Use assertionFailure(_:) instead of assert(false, …)"
  precondition_false:
    name: "Precondition False"
    regex: '(?<!\.)\b(?:Swift\.)?precondition\(false,'
    message: "Use preconditionFailure(_:) instead of precondition(false, …)"
  mark_space:
    name: "Empty line after // MARK"
    match_kinds: 
      - comment
    regex: '// MARK:[^\n]*(?=\n[^\n\S]*\S)'
    message: "Leave an empty line after // MARK"
    severity: warning
  space_after_comment:
    name: "Space after //"
    match_kinds: 
      - comment
    regex: '//[^/\s]'
    message: "Leave a space after //"
    severity: warning
  func_parameter_splitting:
    name: "Function declaration parameter splitting"
    regex: '(?<!\S)(?:func \w{1,30}|init\??)(?:<[^\n(]+>)?\(\s*[^\n,]{1,30}(?:,[\s--\n]*+\n\s*+[^\n,]{1,60})+\s*\)'
    message: "Avoid splitting parameters of a function or initializer in new lines unless the line is very long"
    severity: warning
  image_literal_with_extension:
    name: "Swift image literal with extension"
    regex: '#imageLiteral\(resourceName: "(?:[^\\".]|\.)*\.(?:[^\\"]|\.)*"\)'
    message: "Swift image literals should not have an extension. This is an Xcode bug. Please copy the image literal into another text editor and fix the name"
    severity: error
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
No

* Which Xcode version are you using (check `xcodebuild -version`)?
11.3

* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```bash
echo "array.sorted { \"\\($0)\" < \"\\($1)\" }" | swiftlint lint --no-cache --use-stdin --enable-all-rules
```

**Comment by jpsim** [comment]
Another variant from https://github.com/realm/SwiftLint/issues/3524

```swift
// This triggers a violation:
extension StaticString: Equatable {
    public static func == (lhs: StaticString, rhs: StaticString) -> Bool {
        return "\(lhs)" == "\(rhs)"
    }
}
```
